lib/graphql/schema/validation.rb

```diff
@@ -155,6 +154,17 @@ def self.assert_named_items_are_valid(item_name, get_items_proc)
             # :ok_hand:
           end
         }
+
+        RESERVED_TYPE_NAME = ->(type) {
+          return unless type.name.start_with?('__')
+          return if INTROSPECTION_TYPES[type.name] && INTROSPECTION_TYPES[type.name] == type
+          "Name #{type.name.inspect} must not begin with \"__\", which is reserved by GraphQL introspection."
+        }
+
+        RESERVED_NAME = ->(named_thing) {
```

Comment:
This is a naïve check since there is currently no argument or field name that begins with `__` in the introspection schema. When the day comes, we can build an array of possible values at runtime and compare against that.